clients/src/main/java/org/apache/kafka/common/security/ssl/SslFactory.java

```diff
@@ -419,15 +419,15 @@ void handshake(SslEngineValidator peerValidator) throws SSLException {
             while (true) {
                 switch (handshakeStatus) {
                     case NEED_WRAP:
-                        if (netBuffer.position() != 0) // Wait for peer to consume previously wrapped data
-                            return;
                         handshakeResult = sslEngine.wrap(EMPTY_BUF, netBuffer);
                         switch (handshakeResult.getStatus()) {
                             case OK: break;
                             case BUFFER_OVERFLOW:
                                 netBuffer.compact();
                                 netBuffer = Utils.ensureCapacity(netBuffer, sslEngine.getSession().getPacketBufferSize());
                                 netBuffer.flip();
+                                if (netBuffer.position() != 0) // Wait for peer to consume previously wrapped data
```

Comment:
I think we can do this return at the start of `case BUFFER_OVERFLOW` since we don't need to expand the buffer at this time (we are waiting for peer to consume the data in the buffer by returning).